documentation/codequest.1
.TH man 1 "September 9, 2010"
.LO 1
.SH NAME
codequest \- compute the quantum subsystem code implemented by the
given operators
.SH SYNOPSIS
.B man 
.RB [ \-fhw ]
.RB [ --version ] 
.RB [ -- ]
.RI [ filename ] 

.SH DESCRIPTION
.B codequest
computes a quantum subsystem code that can be implemented by the given measurement
operators.  If you specify
.B -o
then the computed code is optimized, and the optimal qubit distances
are output with the code.  Be aware that the code optimization can be
a very expensive calculation, which is why it is not performed by
default.
.P
In both the optimal and non-optimal case, note that the code that is
output by
.B codequest
is not unique;  in particular, a code with equivalent properties can
be obtained by multiplying the logical and gauge qubit by one of the stabilizers.
.P
.B codequest
accepts one of two different formats,
.I dense
and
.I sparse
for specifying the measurement operators.  See below for a description
of these two formats.
.P
If you specify a filename, then codequest reads the list of
measurement operators from that file.  Otherwise it reads them from
the standard input.
 
.SH OPTIONS
.TP
.BR \-\^f " " input_format " or " \-\-format " " input_format
.RI "Specify the input format, which can be either " dense " or " sparse " as described below.  The default is " dense "."
.TP
.BR \-\^h " or " \-\-help
Print a help message and then exit.
.TP
.BR \-\^o " or " \-\-optimize
Optimize the subsytem code (very expensive!) and output the optimal
logical qubit distances.  Note that this will result in a
different set of logical operators compared to running
.B codequest
without this option. 
.TP
.BR \-\-version
Display the version and then exit.
.TP
.BR \-\- " or " \-\-ignore_rest
Ignores the rest of the labeled arguments following this flag.
.TP
.I [filename]
If
.I filename
is present, then
.B codequest
reads the list of operators from it.  Otherwise, it reads them from
standard input.

.SH "INPUT FORMATS"

.SS DENSE
.P
In the dense format, n-qubit operators are specified in tensor product
form as n-character strings consisting of the letters X, Y, and Z
(case insensitive) to specify a Pauli operator and either spaces or
periods to specify the identity operators.  For example, the
stabilizers of the 5-qubit Kitaev could be specified
as follows:

         XZZX.
        .XZZX
         X.XZZ
         ZX.XZ

.SS SPARSE
.P
In the sparse format, each operator is specified by a line containing
pairs of the form (L) P, where L is a label enclosed in parenthesis
and P is the Pauli operator (i.e., X, Y, or Z, case insensitive)
associated with that label.  The label can be an arbitrary (case
sensitive) string, and will be interpreted as the name of a qubit.
The order of the pairs is not important, though no label may appear
twice.  Not all qubit names need appear in every operator.  If the
name of a qubit does not appear in an operator, then this is
interpreted as meaning that there is an identity operator at that
qubit's location in the tensor product.  Whitespace will be trimmed
around the Pauli operators and around the label, so that (A) and ( A )
are treated as equivalent labels but (AA) and (A A) are not.

For example, the following could be a specification of
operators in a four-qubit system:

         (A) X (B) Y
         (C) Z (A) Z
         (B) X (D) X

The first operator is a tensor product with identities at the
locations of C and D, the X Pauli operator at the location of qubit A,
and the Y Pauli operator at the location of qubit B.  And so on.

If appropriate, the labels will be intepreted as coordinates.
Specifically, labels which are arbitrary integers are interpreted as
one-dimensional coordinates, and labels which are pairs of arbitrary
integers separated by a comma are interpreted as as two-dimensional
coordinates.  The only effect that this has is on how the output is
formatted; the operators of the subsystem code will be displayed using
either one-dimensional or two-dimensional picture representations of
the operators with the physical qubits drawn at locations
corresponding to the coordinates specified for them in the input
operators.

For example, if one inputs the following operators:

         (0,0) X (1,0) X
         (0,0) Z (0,1) Z
         (0,1) X (1,1) X
         (1,0) Z (1,1) Z

Then the X logical qubit of the code is displayed as

        .X
        .X
 
and the Z logical qubit of the code is displayed as

         ZZ
        ..

There need not be a qubit located at every coordinate;  those
coordinates between coordinates where qubits are located will have
spaces drawn, such as follows:

         . Y 
        . Y .
         . . 

If any label cannot be intepreted as a one- or two-dimensional
coordinate, then all labels will be treated as strings instead.
In this case,
.B codequest
places an ordering on the labels and displays the output operators as
one-dimensional strings.  Note that, for example, there is nothing
stopping one for specifying labels in the form of
.IR three -dimensional
coordinates, but they will be interpreted by codequest as string
labels for the purpose of displaying the output.

.SH AUTHOR
Gregory M. Crosswhite is the author of
.BR codequest "."
He can be contacted at
.IR gcross@phys.washington.edu "."


sources/codequest.cc
//@+leo-ver=5-thin
//@+node:gcross.20081119221421.5: * @thin codequest.cc
//@@language c

#include <string>
#include <iostream>
#include <vector>

#include <tclap/CmdLine.h>

using namespace std;
using namespace TCLAP;

extern int main_dense_1d(string filename, bool compute_weights_flag, bool yaml_flag);
extern int main_sparse(string filename, bool compute_weights_flag);

int main(int argc, char** argv) {

try {
	CmdLine cmd("codequest description message", ' ', "1.0");

    vector<string> allowed_formats;
    allowed_formats.push_back("dense");
    allowed_formats.push_back("sparse");
    ValuesConstraint<string> allowedVals( allowed_formats );
    ValueArg<string> formatArg("f","format","Input and output format of quantum operators;  if no format is specified, then defaults to dense.",false,"dense",&allowedVals,cmd);

    SwitchArg weightsSwitch("o","optimize","Optimize the subsytem code (very expensive!) and output the optimal logical qubit distances.  Note that this will result in a different set of logical operators compared to running codequest without this option.", false);
    cmd.add(weightsSwitch);

    SwitchArg yamlSwitch("y","yaml","Print the results in YAML format.  (Only takes effect when using dense input.)", false);
    cmd.add(yamlSwitch);

    UnlabeledValueArg<string> filenameArg("filename","Input file with quantum operators;  if no file is specified, then defaults to standard input.",false,"","filename",cmd);

	cmd.parse( argc, argv );

    if(formatArg.getValue().compare("sparse") == 0) {
        return main_sparse(filenameArg.getValue(),weightsSwitch.getValue());
    } else {
        return main_dense_1d(filenameArg.getValue(),weightsSwitch.getValue(),yamlSwitch.getValue());
    }
} catch (ArgException &e)  // catch any exceptions
	{ cerr << "error: " << e.error() << " for arg " << e.argId() << endl; }
}
//@-leo
